cantools/database/utils.py

```diff
@@ -104,21 +104,17 @@ def decode_data(data: bytes,
 
     unpacked = formats.big_endian.unpack(bytes(data), allow_truncated=allow_truncated)
 
-    if len(data) < expected_length and allow_truncated:
+    if allow_truncated and len(data) < expected_length:
         # to deal with truncated little-endian signals, we have to pad
         # the raw data and remove the spurious signals after unpacking
         # (i.e., before the signal values themselfs get decoded)
-        le_data = bytes(bytearray([0x00]*(expected_length - len(data)))
-                        + data[::-1])
-
+        le_data = data[::-1].rjust(expected_length, b"\xFF")
         le_unpacked = formats.little_endian.unpack(le_data)
 
         # remove spurious little endian signals
+        valid_bit_count = len(data) * 8
         for signal in fields:
-            if signal.byte_order != "little_endian":
```

Comment:
Is this check necessary? All signals in the padded range should be removed, right?

Reply:
So removing the check is fine? le_unpacked cannot contain big endian fields anyway, right?

Reply:
yes and yes. the reason for that if statement might also be due to the history of that patch: if using padding for both little and big endian signals and decoding them into the same dictionary, the condition does not work for BE signals... (I don't remember anymore if that was the case, though.)

be aware that for systems with lots of truncated frames exhibiting little endian signals, reducing the number of required python instructions to be executed as outlined above probably yields a sizeable performance improvement. That said, I think this is an edge case and I consider this solution about as clunky as the current code but slightly harder to read. I'm fine with whatever you consider to be better...